packages/stargate/src/stargateclient.ts

```diff
@@ -95,10 +91,12 @@ export class StargateClient {
    */
   public async getAllBalancesUnverified(address: string): Promise<readonly Coin[]> {
     const path = "/cosmos.bank.Query/AllBalances";
-    const request = QueryAllBalancesRequest.encode({ address: Bech32.decode(address).data }).finish();
+    const request = cosmos.bank.QueryAllBalancesRequest.encode({
```

Comment:
The one open question I have is if this whole method could be auto-generated?

There is some grpc service spec that has the module path (`cosmos.bank.Query`) and say `QueryAllBalancesRequest` maps to `QueryAllBalancesResponse`, along with a method name "AllBalances".

And even if too much of a pain to auto-generate that, it would be good to at least doc where one can find the list of all those types, so you can manually write these functions quickly (go to this file and fill out this pattern). In this example: https://github.com/cosmos/cosmos-sdk/blob/master/proto/cosmos/bank/query.proto#L11-L23

Note there are 4 total queries on bank and we only use 1. (And one we use a provable variant for). It might be cool to auto-gen all these methods as much as practical and then just hand-write the proveable (key) queries. But then again, there is diminishing returns... 10 hours of coding frameworks to save a 20 minute coding task is not worth it.

Reply:
There is some fun stuff autogenerated, which I do not yet fully unserstand. But it might already be almost complete:

```ts
    /** Represents a Query */
    class Query extends $protobuf.rpc.Service {
      /**
       * Constructs a new Query service.
       * @param rpcImpl RPC implementation
       * @param [requestDelimited=false] Whether requests are length-delimited
       * @param [responseDelimited=false] Whether responses are length-delimited
       */
      constructor(rpcImpl: $protobuf.RPCImpl, requestDelimited?: boolean, responseDelimited?: boolean);

      /**
       * Creates new Query service using the specified rpc implementation.
       * @param rpcImpl RPC implementation
       * @param [requestDelimited=false] Whether requests are length-delimited
       * @param [responseDelimited=false] Whether responses are length-delimited
       * @returns RPC service. Useful where requests and/or responses are streamed.
       */
      public static create(
        rpcImpl: $protobuf.RPCImpl,
        requestDelimited?: boolean,
        responseDelimited?: boolean,
      ): Query;

      /**
       * Calls Balance.
       * @param request QueryBalanceRequest message or plain object
       * @param callback Node-style callback called with the error, if any, and QueryBalanceResponse
       */
      public balance(
        request: cosmos.bank.IQueryBalanceRequest,
        callback: cosmos.bank.Query.BalanceCallback,
      ): void;

      /**
       * Calls Balance.
       * @param request QueryBalanceRequest message or plain object
       * @returns Promise
       */
      public balance(request: cosmos.bank.IQueryBalanceRequest): Promise<cosmos.bank.QueryBalanceResponse>;

      /**
       * Calls AllBalances.
       * @param request QueryAllBalancesRequest message or plain object
       * @param callback Node-style callback called with the error, if any, and QueryAllBalancesResponse
       */
      public allBalances(
        request: cosmos.bank.IQueryAllBalancesRequest,
        callback: cosmos.bank.Query.AllBalancesCallback,
      ): void;

      /**
       * Calls AllBalances.
       * @param request QueryAllBalancesRequest message or plain object
       * @returns Promise
       */
      public allBalances(
        request: cosmos.bank.IQueryAllBalancesRequest,
      ): Promise<cosmos.bank.QueryAllBalancesResponse>;

      /**
       * Calls TotalSupply.
       * @param request QueryTotalSupplyRequest message or plain object
       * @param callback Node-style callback called with the error, if any, and QueryTotalSupplyResponse
       */
      public totalSupply(
        request: cosmos.bank.IQueryTotalSupplyRequest,
        callback: cosmos.bank.Query.TotalSupplyCallback,
      ): void;

      /**
       * Calls TotalSupply.
       * @param request QueryTotalSupplyRequest message or plain object
       * @returns Promise
       */
      public totalSupply(
        request: cosmos.bank.IQueryTotalSupplyRequest,
      ): Promise<cosmos.bank.QueryTotalSupplyResponse>;

      /**
       * Calls SupplyOf.
       * @param request QuerySupplyOfRequest message or plain object
       * @param callback Node-style callback called with the error, if any, and QuerySupplyOfResponse
       */
      public supplyOf(
        request: cosmos.bank.IQuerySupplyOfRequest,
        callback: cosmos.bank.Query.SupplyOfCallback,
      ): void;

      /**
       * Calls SupplyOf.
       * @param request QuerySupplyOfRequest message or plain object
       * @returns Promise
       */
      public supplyOf(request: cosmos.bank.IQuerySupplyOfRequest): Promise<cosmos.bank.QuerySupplyOfResponse>;
    }
```

One way or another, we need some query layer between Tendermint and the high level StargateClient.

Reply:
Very interesting class there.

It looks like we could provide an [RPCImpl function](http://protobufjs.github.io/protobuf.js/global.html#RPCImpl): `RPCImpl(method, requestData, callback)` that we can implement with something like `await this.queryUnverified(path, request)` (current code).

With that, we could auto-generate all the grpc endpoints directly in cosmjs that go directly to tendermint rpc.